File: binary_sort.cpp
#include <vector>
#include <iostream>
#include <chrono>

int bin_sort (std::vector <int> v, int b) {
    if (b == v.front()) {
        return 0;
    }
    else if ( b == v.back()) {
        return (v.size() - 1);
    }
    else {
        int bias = 0;
        
        while (v.size() > 1) {
            int half = v.size()/2;

            if (b == v.at(half)) {
                return half+bias;
            }
            else if (b < v.at(half)) {
                v = std::vector <int> (v.begin(), v.end()-half);
                continue;
            } 
            else {
                v = std::vector <int> (v.begin()+half, v.end());
                bias +=half;                
            }
        }
    return -1;
    }
}

int main () {

    std::vector <int> new_vec;
    int to_find, answer;
    // auto start =  std::chrono::high_resolution_clock::now(); // To measure speed of execution
    answer = bin_sort(new_vec, to_find);
    //auto stop =  std::chrono::high_resolution_clock::now(); // To measure speed of execution
    //auto duration = std::chrono::duration_cast <std::chrono::microseconds> (stop - start);
    //std::cout << "Elasped time is " << duration.count() << std::endl;    
    std::cout << "Location of " << to_find << " is " <<  answer << std::endl;

    return 0;
}

File: binary_sort.go
package main

func bin_sort(in_arr []int, to_find int) int64 {
	last_ele := len(in_arr) - 1
	bias := 0
	if to_find == in_arr[0] {
		return 0
	} else if to_find == in_arr[last_ele] {
		return int64(last_ele)
	} else {
		for len(in_arr) > 1 {
			half := len(in_arr) / 2
			if to_find == in_arr[half] {
				return int64(half + bias)
			} else if to_find < in_arr[half] {
				in_arr = in_arr[:half]
				continue
			} else {
				in_arr = in_arr[half:]
				bias += half
			}
		}

	}
	return -1
}

func main() {
	
	start := time.Now()
	res := bin_sort(sample_arr, to_find)
	end := time.Now()

	dura := end.Sub(start)
	fmt.Println("Positon of ", to_find, " is ", res, " took ", dura.Microseconds())


}
